packages/dialog/src/vaadin-dialog.js

```diff
@@ -53,18 +86,200 @@ export class DialogOverlay extends OverlayElement {
   static get properties() {
     return {
       modeless: Boolean,
 
-      withBackdrop: Boolean
+      withBackdrop: Boolean,
+
+      headerTitle: String,
+
+      headerRenderer: Function,
+
+      footerRenderer: Function
     };
   }
 
+  _headerRendererChange(headerRenderer, footerRenderer, opened) {
+    const headerRendererChanged = this.__oldHeaderRenderer !== headerRenderer;
+    this.__oldHeaderRenderer = headerRenderer;
+
+    const footerRendererChanged = this.__oldFooterRenderer !== footerRenderer;
+    this.__oldFooterRenderer = footerRenderer;
+
+    const openedChanged = this._oldOpened !== opened;
+    this._oldOpened = opened;
+
+    if (headerRendererChanged) {
+      const headerContainer = this.headerContainer;
+      headerContainer.innerHTML = '';
+      // Whenever a Lit-based renderer is used, it assigns a Lit part to the node it was rendered into.
+      // When clearing the rendered content, this part needs to be manually disposed of.
+      // Otherwise, using a Lit-based renderer on the same node will throw an exception or render nothing afterward.
+      delete headerContainer._$litPart$;
+    }
+
+    if (footerRendererChanged) {
+      const footerContainer = this.footerContainer;
+      footerContainer.innerHTML = '';
+      // Whenever a Lit-based renderer is used, it assigns a Lit part to the node it was rendered into.
+      // When clearing the rendered content, this part needs to be manually disposed of.
+      // Otherwise, using a Lit-based renderer on the same node will throw an exception or render nothing afterward.
+      delete footerContainer._$litPart$;
+    }
+
+    if (
+      (headerRenderer && (headerRendererChanged || openedChanged)) ||
+      (footerRenderer && (footerRendererChanged || openedChanged))
+    ) {
+      if (opened) {
+        this.requestContentUpdate();
+      }
+    }
+    this._toggleHasHeaderAttribute();
+    this._toggleHasFooterAttribute();
+  }
+
+  _toggleHasHeaderAttribute() {
+    if (this.headerRenderer) {
+      this.setAttribute('has-header', '');
+    } else {
+      this.removeAttribute('has-header');
+    }
+  }
+
+  _toggleHasFooterAttribute() {
+    if (this.footerRenderer) {
+      this.setAttribute('has-footer', '');
+    } else {
+      this.removeAttribute('has-footer');
+    }
+  }
+
+  _headerTitleChanged(headerTitle, opened) {
+    const headerTitleChanged = this._oldHeaderTitle !== headerTitle;
+    this._oldHeaderTitle = headerTitle;
+
+    const openedChanged = this._oldOpened !== opened;
+    this._oldOpened = opened;
+
+    if (headerTitleChanged || openedChanged) {
+      if (this.opened) {
+        this.requestContentUpdate();
+      }
+    }
+    this._toggleHasTitleAttribute();
+  }
+
+  _toggleHasTitleAttribute() {
+    if (this.headerTitle) {
+      this.setAttribute('has-title', '');
+    } else {
+      this.removeAttribute('has-title');
+    }
+  }
+
+  _headerTitleRenderer() {
+    const headerTitleElement = this.getHeaderTitleElement();
```

Comment:
Looks like the title element is always created. We probably should only create it when `headerTitle` is set, otherwise this might break existing `renderer` functions checking for `root.firstChild`.